tutorials/getting_started/creating_a_model.md

```diff
@@ -0,0 +1,197 @@
+  but we can use clever dynamic programming tricks to do so efficiently.
+
+## Implementing the CRF Tagger Model
+
+The `CrfTagger` is not terribly different from the `SimpleTagger` model,
+so we can take that as a starting point. We need to make the following changes:
+
+* define `START_TAG` and `END_TAG` sentinels and make sure they're included
+  in our vocabulary's "labels" namespace
+* give our model a `crf` attribute containing an appropriately initialized
+  `ConditionalRandomField` module
+* create a private `_viterbi_tags()` method that uses the logits from the
+  `tag_projection_layer`, the `transitions` from the CRF module,
+   and the [Viterbi algorithm](https://en.wikipedia.org/wiki/Viterbi_algorithm)
+   to compute the most likely sequence of tags for a given input.
+* replace the softmax class probabilities with the Viterbi-generated most likely tags
+* replace the softmax + cross-entropy loss function
+  with the negative of the CRF log-likelihood
+
+We can then register the new model as `"crf_tagger"`.
+
+## Creating a Dataset Reader
+
+The [CoNLL data](https://www.clips.uantwerpen.be/conll2003/ner/) is formatted like
+
+```
+   U.N.         NNP  I-NP  I-ORG
+   official     NN   I-NP  O
+   Ekeus        NNP  I-NP  I-PER
+   heads        VBZ  I-VP  O
+   for          IN   I-PP  O
+   Baghdad      NNP  I-NP  I-LOC
+   .            .    O     O
+```
+
+where each line contains a token, a part-of-speech tag, a syntactic chunk tag, and a named-entity tag.
+An empty line indicates the end of a sentence, and a line
+
+```
+-DOCSTART- -X- O O
+```
+
+indicates the end of a document. (Our reader is concerned only with sentences
+and doesn't care about documents.)
+
+You can poke at the code yourself, but at a high level we use
+[`itertools.groupby`](https://docs.python.org/3/library/itertools.html#itertools.groupby)
+to chunk our input into groups of either "dividers" or "sentences".
+Then for each sentence we split each row into four columns,
+create a `TextField` for the token, and create a `SequenceLabelField`
+for the tags (which for us will be the NER tags).
+
+## Creating a Config File
+
+As the `CrfTagger` model is quite similar to the `SimpleTagger` model,
+we can get away with a similar configuration file. We need to make only
+a couple of changes:
+
+* change the `model.type` to `"crf_tagger"`
+* change the `"dataset_reader.type"` to `"conll2003"`
+* add a `"dataset_reader.tag_label"` field with value "ner" (to indicate that the NER labels are what we're predicting)
+
+We don't *need* to, but we also make a few other changes
+
+* following [Peters, Ammar, Bhagavatula, and Power 2017](https://www.semanticscholar.org/paper/Semi-supervised-sequence-tagging-with-bidirectiona-Peters-Ammar/73e59cb556351961d1bdd4ab68cbbefc5662a9fc), we use a GRU character encoding
+as well as a GRU for our phrase encoder
+* we also start with pretrained GloVe vectors for our token embeddings
```

Comment:
Link to https://nlp.stanford.edu/projects/glove/?

Reply:
done